### PR TITLE
Add decoding of non-compact uints

### PR DESCRIPTION
**Pull request by ivan4th**
The encoder function were already there.
Also, add tests for non-compact uint encoding/decoding.

### CODE REVIEW

File: decoder_test.go [hunk 1]
```diff
@@ -400,3 +400,34 @@ func TestCompactIntegersBoundaries(t *testing.T) {
 		}
 	})
 }
+
+func decodeTest[T any](t *testing.T, value []byte, expect T) {
+	buf := bytes.NewBuffer(value)
+	dec := NewDecoder(buf)
+	switch typed := any(expect).(type) {
+	case uint8:
```
Comment (poszu) [inline decoder_test.go:408]: It's a weird mix, the function is generic but uses dynamic dispatch to find the type. How about taking `any` instead?
```suggestion
func decodeTest(t *testing.T, value []byte, expect any) {
	buf := bytes.NewBuffer(value)
	dec := NewDecoder(buf)
	switch typed := expect.(type) {
	case uint8:
```

Btw, IMHO this should just be inlined in the test body.

File: decoder.go [hunk 2]
```diff
@@ -93,6 +94,30 @@ func DecodeByte(d *Decoder) (byte, int, error) {
 	return d.scratch[0], n, err
 }
 
+func DecodeUint16(d *Decoder) (uint16, int, error) {
+	n, err := d.read(d.scratch[:2])
```
Comment (fasmat) [inline decoder.go:98]: I think its a bit odd that `scratch` is a field of the decoder. Every `Decode`* function overwrites the contents of it at the beginning anyway, so it doesn't serve as a buffer for bytes that have been read and need to be processed on the next `Decode`* call.

Would it make sense to not have `scratch` be a field of `Decoder` but instead just a local variable? If that causes too many allocations we could use a pool instead. Wdyt?

File: decoder_test.go [hunk 3]
```diff
@@ -20,9 +20,9 @@ func (t *halfReader) Read(buf []byte) (int, error) {
 	return n, nil
 }
 
-// testEncode is a generic encode for primitive types.
+// encodeCompact is a generic encode for primitive types using compact representation.
 // other types are derived from this types.
-func testEncode(tb testing.TB, value any) []byte {
+func encodeCompact(tb testing.TB, value any) []byte {
```
Comment (fasmat) [inline decoder_test.go:25]: NIT (because its just renamed in this PR and not a new function): I don't quite see the value of this generic function. Its a big switch/case statement that does different things based on the value that is passed as `any` to it.

In my opinion the tests in this file would be much easier to follow if they weren't so generic. E.g.

```go
func TestReadFullNew(t *testing.T) {
	t.Run("uint8", func(t *testing.T) {
		val := uint8(math.MaxUint8)
		buf := bytes.NewBuffer(nil)
		enc := NewEncoder(buf)
		_, err := EncodeCompact8(enc, val)
		require.NoError(t, err)

		dec := NewDecoder(buf)
		rst, _, err := DecodeCompact8(dec)
		require.NoError(t, err)
		require.Equal(t, val, rst)
	})
	// other sub tests for other types
}
```

and

```go
func TestDecodeCompactIntegersNew(t *testing.T) {
	for i, tc := range []struct {
		expect uint8
		value  []byte
	}{
		{0, []byte{0b0000_0000}},
		{1, []byte{0b0000_0100}},
		{maxUint6, []byte{0b1111_1100}},
		{maxUint8, []byte{0b1111_1101, 0b0000_0011}},
	} {
		t.Run(fmt.Sprintf("uint8 %d", i), func(t *testing.T) {
			buf := bytes.NewBuffer(tc.value)
			dec := NewDecoder(buf)

			rst, _, err := DecodeCompact8(dec)
			require.NoError(t, err)
			require.Equal(t, tc.expect, rst)
		})
	}
	// next test cases for uint16, uint32, uint64, ...
}
```

Wdyt?

**Review by fasmat — COMMENTED** [review]
I'm not a big fan of how tests in `encoder_test.go` and `decoder_test.go` are structured. They are quite hard to follow because they are so indirect and generic.

I think it would be easier to maintain those tests if they would be more explicit like in my examples above and they would hardly be longer.

File: decoder.go (hunk 2 above)
Comment (ivan4th) [reply decoder.go:98]: I thought about making `scratch` a local variable but there's a possible problem with. Namely, the buffer is passed to `Write` method of an `io.Writer`, and thus it is considered to be "escaping", so that it gets allocated on heap instead of being allocated on the stack. When encoding or decoding a large structure (big slices, etc.), this may easily lead to multitude of small heap allocations, potentially increasing GC pressure and thus harming performance. When reusing the same scratch buffer instead encoder/decoder which is used to encode/decode the whole structure, the extra allocations don't happen b/c the buffer is reused all the time.
I'm not 100% confident that's what exactly will happen, but making such changes needs some careful profiling.

More info regarding stack vs heap alloc: https://medium.com/eureka-engineering/understanding-allocations-in-go-stack-heap-memory-9a2631b5035d

File: decoder_test.go (hunk 1 above)
Comment (ivan4th) [reply decoder_test.go]: Simplified the test (see test-related comment below)

File: decoder_test.go (hunk 3 above)
Comment (ivan4th) [reply decoder_test.go]: I agree that the way the tests are structured obscures their purpose, so I updated the tests for both encoder and decoder.
I didn't completely unroll the tests though, as e.g. 400+ lines of `TestReadFull` repeating the same thing for different `Encode*/Decode*` funcs IMO obscures the purpose of the test and makes actual test case data and encoder/decoder calls near invisible among the boilerplate repetition
.
I made some (hopefully) more clean generic test funcs that are easy to read enough and also, importantly, highlight the convention that the `Encode*/Decode*` funcs must all have the same signature.

**Review by poszu — APPROVED** [review]

File: encoder_test.go [hunk 4]
```diff
@@ -39,22 +39,22 @@ func BenchmarkEncodeStrings_WithWriterForStrings(b *testing.B) {
 	}
 }
 
-type compactTestCase[T any] struct {
+type encTestCase[T any] struct {
```
Comment (fasmat) [inline encoder_test.go:42]: NIT: I liked the previous name better, because the examples here are used for both encoding and decoding tests 🙂

**Review by fasmat — APPROVED** [review]

File: encoder_test.go (hunk 4 above)
Comment (ivan4th) [reply encoder_test.go]: Fixed, thanks
